web/bitwarden/bitwarden.go

```diff
@@ -264,6 +272,52 @@ func getInitialCredentials(c echo.Context) error {
 	return c.JSON(http.StatusOK, out)
 }
 
+// checkTwoFactor returns true if the request has a valid 2FA code.
+func checkTwoFactor(c echo.Context, inst *instance.Instance) bool {
+	cache := config.GetConfig().CacheStorage
+	key := "bw-2fa:" + inst.Domain
+
+	if passcode := c.FormValue("twoFactorToken"); passcode != "" {
+		if token, ok := cache.Get(key); ok {
+			if inst.ValidateTwoFactorPasscode(token, passcode) {
+				return true
+			}
+		}
+	}
+
+	email, err := inst.SettingsEMail()
+	if err != nil {
+		_ = c.JSON(http.StatusInternalServerError, echo.Map{
+			"error": err.Error(),
+		})
+		return false
+	}
+	var obscured string
+	if parts := strings.SplitN(email, "@", 2); len(parts) == 2 {
+		s := strings.Map(func(_ rune) rune { return '*' }, parts[0])
+		obscured = s + "@" + parts[1]
+	}
+
+	token, err := lifecycle.SendTwoFactorPasscode(inst)
+	if err != nil {
+		_ = c.JSON(http.StatusInternalServerError, echo.Map{
+			"error": err.Error(),
+		})
+		return false
+	}
+	cache.Set(key, token, 5*time.Minute)
+
+	_ = c.JSON(http.StatusBadRequest, echo.Map{
```

Comment:
I don't understand why we build those HTTP responses without sending them.

Reply:
The HTTP responses are sent. `c.JSON` will really write in `c.Response`. We just don't check if there is an error, because on both cases, we are doing the same.

Reply:
OK, great. Thank you.